doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -689,198 +734,548 @@ one from Ltac1, except that it requires the goal to be focused.
+
+   .. example::
+
+      Assume we perform:
+
+      .. coqdoc::
+
+         Ltac2 Notation "foo" c(thunk(constr)) ids(list0(ident)) := Bar.f c ids.
+
+      Then the following expression
+
+      `let y := @X in foo (nat -> nat) x $y`
+
+      will expand at parsing time to
+
+      `let y := @X in`
+      `let c := fun () => constr:(nat -> nat) with ids := [@x; y] in Bar.f c ids`
+
+      Beware that the order of evaluation of multiple let-bindings is not specified,
+      so that you may have to resort to thunking to ensure that side-effects are
+      performed at the right time.
+
+Abbreviations
+~~~~~~~~~~~~~
+
+.. cmd:: Ltac2 Notation {| @string | @lident } := @tac2expr6
+   :name: Ltac2 Notation (abbreviation)
 
-  + if :n:`@ltac2_scope` parses :n:`@quotentry`, parses :n:`{? @quotentry}` and produces either :n:`None` or
-    :n:`Some x` where :n:`x` is the parsed expression.
+   Introduces a special kind of notation, called an abbreviation,
+   that does not add any parsing rules. It is similar in
+   spirit to Coq abbreviations (see :cmd:`Notation (abbreviation)`,
+   insofar as its main purpose is to give an
+   absolute name to a piece of pure syntax, which can be transparently referred to
+   by this name as if it were a proper definition.
 
-- :n:`self`:
+   The abbreviation can then be manipulated just like a normal Ltac2 definition,
+   except that it is expanded at internalization time into the given expression.
+   Furthermore, in order to make this kind of construction useful in practice in
+   an effectful language such as Ltac2, any syntactic argument to an abbreviation
+   is thunked on-the-fly during its expansion.
+
+   For instance, suppose that we define the following.
+
+   :n:`Ltac2 Notation foo := fun x => x ().`
+
+   Then we have the following expansion at internalization time.
+
+   :n:`foo 0 ↦ (fun x => x ()) (fun _ => 0)`
+
+   Note that abbreviations are not type checked at all, and may result in typing
+   errors after expansion.
+
+.. _syntactic_classes:
+
+Syntactic classes
+~~~~~~~~~~~~~~~~~
 
-  + parses a Ltac2 expression at the current level and returns it as is.
+The simplest syntactic classes in Ltac2 notations represent individual nonterminals
+from the |Coq| grammar.  Only a few selected nonterminals are available as syntactic classes.
+In addition, there are metasyntactic operations for describing
+more complex syntax, such as making an item optional or representing a list of items.
+When parsing, each syntactic class expression returns a value that's bound to a name in the
+notation definition.
 
-- :n:`next`:
+Metasyntactic classes are described with a form of S-expression:
 
-  + parses a Ltac2 expression at the next level and returns it as is.
+   .. insertprodn ltac2_scope syn_node
 
-- :n:`tactic(n = @int)`:
+   .. prodn::
+      ltac2_scope ::= @string
+      | @int
+      | @syn_node
+      | @syn_node ( {+, @ltac2_scope } )
+      syn_node ::= _
+      | @ident
```

Comment:
Any reason to define `syn_node` instead of reusing the common `name`?

Reply:
It's from the Ltac2 grammar.  I will change.